src/content/en/fundamentals/engage-and-retain/push-notifications/handling-messages.markdown

```diff
@@ -0,0 +1,448 @@
+  event.waitUntil(() => {
+    if (event.data) {
+      return Promise.resolve(event.data);
+    }
+    return fetch('some/data/endpoint.json')
+      .then(response => response.json());
+  });
+});
+{% endhighlight %}
+
+In both cases we end up with a JSON object. Now it's time to show a
+notification to the user.
+
+{% highlight javascript %}
+self.addEventListener('push', event => {
+  event.waitUntil(() => {
+    if (event.data) {
+      return Promise.resolve(event.data);
+    }
+    return fetch('some/data/endpoint.json')
+        .then(response => response.json());
+    })
+    .then(data => {
+      return self.registration.showNotification(data.title, {
+        body: data.body,
+        icon: (data.icon ? data.icon : '/images/icon-192x192.png'),
+        vibrate: [200, 100, 200, 100, 200, 100, 400],
+        tag: data.tag
+       })
+    }))
+});
+{% endhighlight %}
+
+## Combine similar notifications {#combine-similar-notes}
+
+Sometimes it's useful to combine multiple notifications into a single one. For
+example, a social networking app might want to avoid messaging users for every
+post from a particular person, and instead combine them.
+
+<img src="images/combined-notes-mobile.png" width="296" height="526">
+
+Combining similar notifications comes down to three things.
+
+* A call to `getNotifications()`.
+* reusing an existing `tag` value.
+* Setting the `retnotify` flag in the call to `showNotification()`.
+
+Let's look at an example that shows all three.
+
+The following examples, we're going to assume that you've already received or
+retrieved message data  as described in the last section. Now let's look at what
+to do with it. Start with a basic push event handler.
+
+{% highlight javascript %}
+self.addEventListener('push', function(event) {
+  event.waitUntil(
+    // Get the message data somehow and return a Promise.
+  )
+  .then(data => {
+    // Do something with the data.
+  })
+});
+{% endhighlight %}
+
+Check for notifications that match `data.tag` with a call to `getNotifications()`.
+
+{% highlight javascript %}
+self.addEventListener('push', function(event) {
+  event.waitUntil(
+    // Get the message data somehow and return a Promise.
+  )
+  .then(data => {
+    return self.registration.getNotifications({tag: data.tag});
+  })
+  .then(notifications => {
+    //Do something with the notifications.
+  })
+});
+{% endhighlight %}
+
+In other examples, we've instantiated our notification options right in the call 
+to `showNotification()`. For this scenario, the options object needs to change 
+based on the results of `getNotifications()`.  So instantiate a notification 
+options object. 
+
+Notice that we've also attached the notification data to the notification
+options. We're doing this to ensure that it's available to `notificationclick`,
+which we'll look at in a later section. To tell the browser we're combining
+notifications, we need to not just reuse the tag (line 6), we also need to set renotify
```

Comment:
i'd simplify this down a bit: To tell the browser we're combining notifications, we need to reuse the tag (line 6) and also set `renotify` to true (line 17)." It takes out the double negative.